docker/lwua-ingest/lwua-py/lwua/watcher.py

```diff
@@ -0,0 +1,84 @@
+import time
+from datetime import datetime
+import os
+from pathlib import Path
+from dotenv import load_dotenv
+from abc import ABC, abstractmethod
+import logging
+from lwua.helpers import enable_logging, resolve_path
+from lwua.graphdb import (
+    ingest_data_file,
+    delete_data_file,
+    get_registry_of_lastmod,
+)
+
+log = logging.getLogger(__name__)
+
+class FolderChangeObserver(ABC):
+    @abstractmethod
+    def added(self, fname: str, lastmod: datetime = None):
+        pass
+
+    @abstractmethod
+    def removed(self, fname: str):
+        pass
+
+    @abstractmethod
+    def changed(self, fname: str, lastmod: datetime = None):
+        pass
+
+class FolderChangeDetector:
+    def __init__(self, folder_to_inspect):
+        self.root = Path(folder_to_inspect)
+        while not self.root.exists():
+            log.info(f"Waiting for {self.root} to exist")
+            time.sleep(1)
+        log.info(f"Watching {self.root}")
+
+    def report_changes(self, known_lastmod_by_fname: dict = {}, observer: FolderChangeObserver = None):
+        current_lastmod_by_fname = {p: datetime.utcfromtimestamp(os.path.getmtime(p)) for p in self.root.glob('**/*') if p.is_file()}
+        log.info(f"current_lastmod_by_fname: {current_lastmod_by_fname}")
+        for fname in known_lastmod_by_fname:
```

Comment:
either do not accept Observer to be None
or guard these lines up until the return statement with an `if observer is not None:`

Reply:
looking at it now it still feels like an `assert observer is not None, "...msg here..."` at the top of this methid will allow for a more helpful message at runtime than the classic `Nonetype has no method xyz...`